Add option to generate function body to goto-instrument

hannes-steffenhagen-diffblue opened the pull request:
This adds the commandline parameters `--replace-function-body` and `--replace-function-body-options`. The first gets a regex that defines a set of functions whose body should be replaced; Whereas the latter specifies what kind of body should be generated. The options here are:

* `nondet-return` same as `havoc` without further options
* `assert-false` generate a single `assert(false);` in the body.
* `assume-false` generate a single `assume(false);` in the body (this has a similar effect to `--undefined-function-is-assume-false`)
* `havoc[,params][,globals:<regex>]` which generates a `return nondet` for non-`void` functions and assigns `nondet` to the targets of pointers to non-`const` if `params` is set and assignments of nondet to non-`const` globals matching the regex if it is set.

Other than that, I'm looking for comments on the command line parameters (I've put the function "goals" and the options for the body replacement in separate options for now, but it feels a bit wordy).

hannes-steffenhagen-diffblue committed:
Add option to generate function body to goto-instrument

tautschnig commented:
May I start with a few questions:
* I believe the existing `remove-function-body` should either be removed and instead yield a meaningful error message or be made to redirect to `replace-function-body+havoc`.
* What is the relationship to #1566?
* What is the relationship to #1585?

hannes-steffenhagen-diffblue commented:
@tautschnig 

* Can you elaborate? One ends with the function having no body, the other ends with the function having a body with (possibly) a `return nondet;` which are "distinguishable" states. Are you saying these should be the same somehow?

* #1566 Replaces calls to a certain function by `assume_false`; This behaves similarly to undefined-function-is-assume false but with a certain target function rather than with all undefined functions. This somewhat overlaps with `--replace -function-body-options assume-false`: The net effect for a "full" goto binary is roughly the same I think, but there are still some differences (for example, you can use replace-function-body on just the goto binary containing the function to replace and then link, whereas you'd need to use make-function-assume-false on all goto-binaries containing the function calls or call it after linking; It also needs to look at all instructions whereas replace-function-body just needs to iterate over the functions, so there might be a slight performance advantage there, though I don't expect it to be particularly large).

* #1585 behaves similarly to `--replace-function-body-options havoc,params`; The difference being that this PR does the "nondetting" of parameters by actually replacing the function body, whereas in the other PR this is done during abstract interpretation (I'm told actually symex) (and thus needs to be implemented separately for each domain). In that PR @peterschrammel suggested doing this through a goto transformation which this PR is attempting to do. Currently it's only doing the simplest approximation of this, which is just directly assigning nondet to the pointer targets; Some other things that are "thinkable" are recursively assigning nondet values to non-null pointers, struct members and array elements (for constant size arrays), but I wanted to know what others think about this before going too crazy on it.

hannes-steffenhagen-diffblue commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1522,8 +1533,13 @@ void goto_instrument_parse_optionst::help()
     " --splice-call caller,callee  prepends a call to callee in the body of caller\n"  // NOLINT(*)
     // NOLINTNEXTLINE(whitespace/line_length)
-    " --undefined-function-is-assume-false\n" // NOLINTNEXTLINE(whitespace/line_length)
-    "                              convert each call to an undefined function to assume(false)\n"
```
Removed text that should still be there

hannes-steffenhagen-diffblue commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,310 @@
+  {
+    function.body.clear();
+    generate_parameter_names(function, symbol_table, function_name);
+    replace_function_body_impl(function, symbol_table, function_name);
```
Should be "generate_function_body", because the body has already been removed here

hannes-steffenhagen-diffblue commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,310 @@
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) override
+  {
```
Put a skip instruction here or rename this to empty

hannes-steffenhagen-diffblue commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,310 @@
+public:
+  havoc_replace_function_bodiest(
+    std::vector<irep_idt> globals_to_havoc,
+    bool havoc_parameters)
```
Use regex for parameter instead of all-in boolean parameter

hannes-steffenhagen-diffblue reviewed (commented):
Add a test for function name regex

Add a test for replace-function-body-options skip

Add an option for assert-false-then-assume-false

hannes-steffenhagen-diffblue commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,310 @@
+  bool havoc_parameters;
+};
+
+std::unique_ptr<replace_function_bodiest> parse_replace_function_body_options(
```
Rename this to "factory..."

hannes-steffenhagen-diffblue commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,310 @@
+  const std::regex &functions_regex,
+  const std::string &replace_options)
+{
+  messaget messages;
```
pass this as parameter instead

hannes-steffenhagen-diffblue commented on src/goto-instrument/replace_function_bodies.h:
```diff
@@ -0,0 +1,27 @@
+
+Module: Replace bodies of goto functions
+
+Author: DiffBlue Limited. All rights reserved.
```
Don't use "All rights reserved"

martin-cs commented:
For the record, the comments above are a composite of mine and @chrisr-diffblue 's reviews of this done sat with @hannes-steffenhagen-diffblue .  Beyond these we are happy to merge this.

martin-cs commented:
There are questions of condensing other options and other functionality.  This patch is the first step of a not-as-covert-as-it-was-before-I-started-writing-this attempt to remove the special case code for functions without bodies from the various analyses.  In line with this I think we should also:

1. Redirect --remove-function-body as @tautschnig suggests.
2. Redirect (and update the description of) --undefined-functions-are-assume-false so that it instead generates function bodies with assume-false in them.

@tautschnig @peterschrammel opinions?

Also @polgreen does the functionality in this PR cover the use cases you had for #1566 and #1585 ?

martin-cs commented:
@hannes-steffenhagen-diffblue : one thing I forgot when we were writing up the review, can we have an option --generate-function-bodies which is --replace-function-body for all functions which don't have a body at the moment.

hannes-steffenhagen-diffblue commented:
@martin-cs I think that's all the comments addressed, except for the option merging thing (maybe separate PR?). Mind taking another look?

thk123 commented on regression/goto-instrument/generate-function-body/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^VERIFICATION FAILED$
+^\[main.assertion.1\] assertion does_not_get_reached: SUCCESS$
+^\[should_be_generated.assertion.1\] assertion FALSE: FAILURE$
```
Missing new line

thk123 commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1448,6 +1449,31 @@ void goto_instrument_parse_optionst::instrument_goto_program()
+    std::regex function_regex = std::regex(
+      only_generate ? cmdline.get_value("generate-function-body")
+                    : cmdline.get_value("replace-function-body"));
+    status() << "Replacing function bodies" << eom;
```
Should this method say `"Generating function bodies"` if `cmdline.get_value("generate-function-body")`

thk123 commented on src/goto-instrument/goto_instrument_parse_options.h:
```diff
@@ -86,7 +86,11 @@ Author: Daniel Kroening, kroening@kroening.com
   "(splice-call):" \
-  OPT_REMOVE_CALLS_NO_BODY
+  OPT_REMOVE_CALLS_NO_BODY \
+  "(replace-function-body):" \
```
But these options and their documentation in to `replace_function_bodies.h` to make it easier for it to be reused in different front ends.

hannes-steffenhagen-diffblue replied:
As a define like some of the other options?

thk123 commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1448,6 +1449,31 @@ void goto_instrument_parse_optionst::instrument_goto_program()
+    cmdline.isset("replace-function-body") ||
+    cmdline.isset("generate-function-body"))
+  {
+    if(
```
I have a mild preference to remove this code from here and put into the constructor of `replace_function_bodiest` so that the logic of how these parameters are handled is separated from the specific binary being used.

hannes-steffenhagen-diffblue replied:
I was actually the opposite; I had considered doing the parsing of the arguments outside of replace_function_bodiest entirely to avoid passing more "meaningful" strings around than we really have to

thk123 replied:
My reason was that if someone wishes to re-use this they must copy/paste the interface to provide a consistent experience for the user. Rather than passing raw strings, what if you pass the whole `cmdlinet` into it? So it is clear they aren't meaningful strings but really the user input. 

The down side to my suggestion is becomes a class that is both responsible for generating functions and parsing command line args. Perhaps it would be better to have a third place that takes a cmdline and returns a `replace_goto_functionst`, but OK with leaving that to the first person who wishes to re-use this code.

tautschnig replied:
I'd like to support @thk123's position: In order to achieve what #1949 says, we will need this in all front-ends. So we really don't want to reimplement the command-line parsing multiple times. In my opinion, the last suggestion to have "a third place that takes a cmdline and returns..." is the best approach.

peterschrammel replied:
Pushing the `only_generate` and `function_regex` logic into a reusable helper in `replace_function_bodies` would be preferable.

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+  throw replace_function_body_parse_error("Can't parse \"", options, "\"");
+}
+
+void replace_function_bodies(
```
As the interface to this whole functionality, could you add documentation here.

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+    {
+      if(
+        parameter.type().id() == ID_pointer &&
+        !is_constant_or_has_constant_components(parameter.type(), ns))
```
Why is this relevant?

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+        {
+          const std::string param_base_name =
+            parameter.get_base_name().empty()
+              ? "__param$" + std::to_string(param_counter++)
```
To avoid collision you might consider using the cprover prefix (I suppose this doesn't matter as `$` is illegal in C variable names, and I think is considered reserved in Java).

hannes-steffenhagen-diffblue replied:
According to the [oracle docs](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html) '$' isn't reserved, but they do strongly recommend not using it in user names. $ is not permitted by the C standard, but GCC and Clang support it as an extension (apparently because some older compilers did), again they recommend against using it.

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+          new_param_sym.name = new_param_identifier;
+          new_param_sym.type = parameter.type();
+          new_param_sym.base_name = param_base_name;
+          new_param_sym.is_parameter = true;
```
Implied by the fact it is a `parameter_symbolt`

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+          new_param_sym.base_name = param_base_name;
+          new_param_sym.is_parameter = true;
+          new_param_sym.mode = ID_C;
+          const symbolt *function_symbol = symbol_table.lookup(function_name);
```
Since you immediately dereference this either INVARIANT it being non-null or use `lookup_ref`

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+          new_param_sym.location = function_symbol->location;
+          symbol_table.add(new_param_sym);
+        }
+        auto function_symbol = *symbol_table.lookup(function_name);
```
Immediate dereference so prefer `lookup_ref`

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+          new_param_sym.location = function_symbol->location;
+          symbol_table.add(new_param_sym);
+        }
+        auto function_symbol = *symbol_table.lookup(function_name);
```
Also what is going on here? Should this not be done after updating all the parameters? Also I think if you want to update the symbol table you should be using `get_writeable` rather than removing and adding an entry

hannes-steffenhagen-diffblue replied:
Yes it should, this was a braces-derp on my end

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+  }
+};
+
+class assume_false_replace_function_bodiest : public replace_function_bodiest
```
Consider documenting each of these classes with:
 a) the arugments needed to use them
 b) a C style code block of what will be produced

thk123 replied:
You might also consider writing some really simple unit tests for each of these classes to verify the program produced.

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+    {
+      if(
+        parameter.type().id() == ID_pointer &&
+        !parameter.type().subtype().get_bool(ID_C_constant) &&
```
Not sure, but perhaps this is the same condition as `is_constant_or_has_constant_components` (e.g. is pointer to a pointer to a constant?)

hannes-steffenhagen-diffblue replied:
I'm not sure; Actually it's also possible that this is the better version; i.e. you can absolutely assign a pointer-to-pointer-to-constant. Then again, structs with constant members usually aren't assignable so I'm not 100% clear on what the correct way to do this here is

thk123 replied:
Consider writing unit tests to figure out - I think it should be clear what is "correct" when you do this

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+};
+
+static std::vector<std::string>
+tokenize(const std::string &base, char separator)
```
Document? Also surely this code exists somewhere already? Is it not `split_string` in `string_utils.cpp`?

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+}
+
+static void write_to_stream(std::ostream &out)
+{
```
huh?

hannes-steffenhagen-diffblue replied:
This is just to have a convenient method to assemble a string... Didn't end up actually needed it here so I'm removing it, but I figure something along those lines might be useful for string-utils?

thk123 replied:
Maybe? I'll reserve judgement until I see a PR with it being used :stuck_out_tongue:

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,367 @@
+      const std::vector<std::string> key_value_pair = tokenize(tokens[i], ':');
+      if(key_value_pair.size() != 2)
+      {
+        throw replace_function_body_parse_error(
```
Avoid introducing `throw std::string` type statements, consider having a `replace_function_body_exceptiont` for this

hannes-steffenhagen-diffblue replied:
Should it be a child of std::runtime_error? If yes, should the message include `__FILE__` and `__LINE__`?

thk123 replied:
They both sound like sensible suggestions to me.

thk123 reviewed (commented):
Not reviewed tests

hannes-steffenhagen-diffblue commented:
@thk123 I think I've addressed most of your comments, mind having another look?

thk123 commented on src/goto-instrument/replace_function_bodies.h:
```diff
@@ -0,0 +1,87 @@
+#include <util/message.h>
+
+/// Replace function bodies with some default behavior
+/// Can generate the following types of function bodies:
```
Explain the lhs of this list is the command line argument

thk123 commented on src/goto-instrument/replace_function_bodies.h:
```diff
@@ -0,0 +1,87 @@
+///   return nondet_int();
+/// }
+///
+/// nondet-return return nondet for non-void functions
```
Nit: `nondet-return:  return nondet for non-void functions` (missing colon separating command line from result)

thk123 commented on src/goto-instrument/replace_function_bodies.h:
```diff
@@ -0,0 +1,87 @@
+/// @param functions_regex Replace bodies of functions whose names matches regex
+/// @param only_generate When true, skip functions that already have a body
+/// @param replace_options Options string indicating what body to generate
+/// @param messages Destination for status/warning messages
```
`\param` also this documentation should go in the cpp next to the implementation rather than the header (as per coding guidelines)

thk123 replied:
I also think it would be more useful to split each up to go with the corresponding class but that might just be personal preference

thk123 commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,350 @@
+        new_param_sym.type = parameter.type();
+        new_param_sym.base_name = param_base_name;
+        new_param_sym.mode = ID_C;
+        auto &function_symbol = symbol_table.lookup_ref(function_name);
```
I think it is clearer if you make the `const` here explicit.

tautschnig commented:
> Also @polgreen does the functionality in this PR cover the use cases you had for #1566 and #1585 ?

@hannes-steffenhagen-diffblue, @polgreen (and possibly @martin-cs, @chrisr-diffblue) have you had a chance to talk/sit down together about this? I really really really think we should not have multiple, competing PRs open without decent communication.

polgreen commented:
I haven't looked at this PR yet. I will do. Does this address the comment Peter made on https://github.com/diffblue/cbmc/pull/1585, i.e., does it create a unifying mechanism for doing things with functions without bodies?


> There are at least four different things what to do with functions without bodies:

    assume(false)
    nondeterministic return value only, no side-effects (current default behaviour)
    nondeterministic return value and side-effects
    use summary

>There should be a unified mechanism for choosing between these. Doing it through a goto-program transformation is the obvious candidate.

>I suggest creating stub bodies for functions without bodies and filling the bodies with the required goto instructions dependent on the chosen function-without-body-replacement-action 1-4.
This would also seamlessly integrate with lazy loading, I think (replacement-action 0 "load/use actual body if available")."

If it does, then I can re-work my changes around this PR, eventually.

tautschnig commented:
To elaborate as requested, though it seems this might have been resolved in the meantime:
> Can you elaborate? One ends with the function having no body, the other ends with the function having a body with (possibly) a `return nondet;` which are "distinguishable" states. Are you saying these should be the same somehow?

They are indeed distinguishable, but they are only sometimes actually treated differently. Which actually is really bad. With your work here I am tempted to suggest that we should forbid the existence of functions without bodies in any analysis pass. Thus, before starting any analysis, missing functions should get replaced by stubs. Any code currently testing "does this function have a body" or "does this function exist in the function map" should have those replaced by invariants.

polgreen commented:
Actually, briefly looking at this PR, it looks like it reimplements the stuff I did over the summer.

tautschnig commented:
> Actually, briefly looking at this PR, it looks like it reimplements the stuff I did over the summer.

I believe so, but it would be really nice if the fact that people are in the same city could be taken advantage of :-)

martin-cs commented:
@hannes-steffenhagen-diffblue : one thing I forgot when we were writing up the review, can we have an option --generate-function-bodies which is --replace-function-body for all functions which don't have a body at the moment.

martin-cs commented:
@polgreen @tautschnig   The genesis of this work was looking at what Elizabeth had done and PRd and trying to apply the architectural comments that Peter and I had on  #1585 (specifically that it should be done by generating bodies, not during analysis).  My impression is that this gives the functionality she wants without the duplication in different analyzers.

@polgreen : If this got merged would you be happy and not feel that it had clashed with your work?

Also, @tautschnig : I agree that a longer-term goal should be to do away with "functions without bodies" in analysis.  I think this is the necessary first step in this direction.

hannes-steffenhagen-diffblue commented:
@martin-cs The current version has `--generate-function-body` with this functionality

martin-cs commented:
@hannes-steffenhagen-diffblue Sorry : race condition on comments and changes.  Thank you.

tautschnig commented on src/goto-instrument/replace_function_bodies.h:
```diff
@@ -0,0 +1,87 @@
+///   __CPROVER_assert(false);
+///   __CPROVER_assume(false); }
+///
+/// havoc[,params:<regex>][,globals:<regex>]:
```
`<regex>` needs some kind of escaping to make Doxygen happy.

hannes-steffenhagen-diffblue replied:
I think I'll just remove the angle brackets rather than escape, for sake of legibility in code

tautschnig commented:
@hannes-steffenhagen-diffblue Let me know when this is deemed final, then I'll review.

hannes-steffenhagen-diffblue commented:
@tautschnig @thk123 

Fixed the doxygen and added the `const` as per @thk123 suggestion. Could you please take another look and see if there's something in here that should be fixed in this PR?

peterschrammel commented:
> With your work here I am tempted to suggest that we should forbid the existence of functions without bodies in any analysis pass. 

Yes, @tautschnig, that's the intent.

> Any code currently testing "does this function have a body" or "does this function exist in the function map" should have those replaced by invariants.

Please raise a ticket for this.

thk123 commented on regression/goto-instrument/replace-function-body-assert-false-assume-false/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^VERIFICATION FAILED$
+^\[main.assertion.1\] assertion 0: SUCCESS$
+^\[crashes_program.assertion.1\] assertion FALSE: FAILURE$
```
Missing new line

thk123 reviewed (commented):
Code changes look fine and the tests are very precise that I don't think there is a need for unit tests. 

I would like to see some more tests with pointers to pointers to constants (i.e points to a pointer whose value can be modified (and therefore should when havocing, but the value it points to should not). You may find the tests named with `fp` in `regression/goto-analyzer` useful, also the code written for that: `analyses/remove_const_function_pointerst` that can be maybe generalised to all potentially constant pointers. 

I don't know if these should be considered blocking - without the tests I think there is a reasonable chance that these tricky edge cases will either be wrong now or will get broken in the future, but it is a new feature so it certainly isn't a regression!

hannes-steffenhagen-diffblue commented:
I've squashed the "address review comments" commits and created a new Github issue for adding more regression tests to this, I'll be working on that this sprint ideally.

martin-cs reviewed (approved):
@chrisr-diffblue and I reviewed this in person.  I believe all of the fixes requested have been made.

chrisr-diffblue reviewed (approved):
Confirming that comments from my offline-review with @martin-cs and @hannes-steffenhagen-diffblue have been addressed.

tautschnig commented:
Future changes/cleanup of analyses now tracked in #1949.

tautschnig commented:
@hannes-steffenhagen-diffblue @polgreen Could you please work together to see whether #1566 and #1585 can safely be closed once this is merged?

tautschnig commented on regression/goto-instrument/generate-function-body/test.desc:
```diff
@@ -0,0 +1,8 @@
+CORE
+main.c
+--generate-function-body '(?!__).*' --replace-function-body-options assert-false
```
Would it be possible to add another *.desc file to this same test directory where different `replace-function-body-options` are specified? This would nicely contrast the behaviour caused by different options on the very same test example.

tautschnig commented on src/goto-instrument/Makefile:
```diff
@@ -61,6 +61,7 @@ SRC = accelerate/accelerate.cpp \
       undefined_functions.cpp \
       uninitialized.cpp \
       unwind.cpp \
+      replace_function_bodies.cpp \
```
I'd suggest to put this in goto-programs (we will need this in many tools).

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.h:
```diff
@@ -86,7 +88,11 @@ Author: Daniel Kroening, kroening@kroening.com
   "(undefined-function-is-assume-false)" \
   "(remove-function-body):"\
```
I think both of the above should be marked as deprecated and should be moved into some macro defined in replace_function_bodies.h.

peterschrammel commented:
@hannes-steffenhagen-diffblue, one of the jobs has a test failure.

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1448,6 +1449,32 @@ void goto_instrument_parse_optionst::instrument_goto_program()
+      cmdline.isset("generate-function-body"))
+    {
+      throw "can only use one of --replace-function-body "
+        "or --generate-function-body";
```
Why are these mutually exclusive? To me, it seems ok to request generating bodies for those that never had one in the first place, and also add some more to that list (of bodiless functions) via replace-function-body.

hannes-steffenhagen-diffblue replied:
I don't like changing how replace-function-body works depending on whether or not there had also been a generate-function-body in the same command line. But if you want something like this maybe I need to rethink the way I currently do the command line argument parsing, could you tell me how you would want to use this from the command line?

tautschnig replied:
I thought that "replace-function-body" operates on those functions that previously had a function body, while "generate-function-body" will work on those that don't. Somehow I must be getting this wrong, as these are disjoint sets and thus there wouldn't be any interaction. Would you mind clarifying?

hannes-steffenhagen-diffblue replied:
@tautschnig `replace-function-body` does "replace" the body of the function with what's specified in `replace-function-body-options`, regardless of whether or not the function has a body. Generate works the same as replace but skips over functions that already have a body.

I did it like that because I thought there might be reasons for why you'd want to only generate bodies for functions that don't have a body yet, but I couldn't think of a reason for why you'd want to replace function  bodies of functions that do already have a body but at the same time leave bodyless functions alone.

I could make these completely separate like you suggested, would that be preferable?

tautschnig replied:
I think subtle interaction is dangerous (and hence you disallowed it). How about making the steps very explicit by leaving in place/promoting `remove-function-body`, dropping `replace-function-body`, and leaving all remaining work to `generate-function-body` (including all the options currently specifiable via `replace-function-body-options`)?

Using `remove-function-body` together with `generate-function-body` would then be perfectly permissible, with the sequence 1) remove and 2) generate.

hannes-steffenhagen-diffblue replied:
@martin-cs @peterschrammel Do you agree that this is a good way to do this? I'm not opposed to doing it this way, but if I'm going to be changing the command line interface I figure it'd be better to have some agreement of how this should work. This would also mean removing the "empty" option which removes bodies (and conflicted with remove-function-body before)

tautschnig replied:
\<marketing\>We would thus not have any surprises/disruption for existing users of `remove-function-body`.\</marketing\>

chrisr-diffblue replied:
Coming a bit late to this discussions, so apologies for that... but I'd agree with @tautschnig that subsuming the new functionality in `replace-function-body` into `generate-function-body` and then just have the users do the `remove-function-body` + ` generate-function-body` sounds like the most predictable and generic way to do this. I like my tools not to surprise me with subtle interactions :-) Presumably if we *really* then wanted a `replace-function-body` convenience option, that could just map on to the remove+generate options as a bit of command line syntactic sugar, but I'd lean towards not adding more options than we absolutely need to....

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+
+Module: Replace bodies of goto functions
+
+Author: DiffBlue Limited.
```
Inconsistent with #1919

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+#include <sstream>
+#include <util/make_unique.h>
+#include <util/message.h>
+#include <langapi/language_util.h>
```
Do you actually need this?

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+  void replace_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name)
```
Pass by (const) reference.

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+  void generate_parameter_names(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) const
```
Pass by (const) reference.

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+    symbol_tablet &symbol_table,
+    irep_idt function_name) const
+  {
+    namespacet ns(symbol_table);
```
`const`

martin-cs commented:
I don't know much about Travis but is this one of the weird transient failures it gives?

```The following tests FAILED:
	 68 - goto-gcc-CORE (Failed)
Errors while running CTest
/Users/travis/.travis/job_stages: line 57: shell_session_update: command not found```

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+        new_param_sym.name = new_param_identifier;
+        new_param_sym.type = parameter.type();
+        new_param_sym.base_name = param_base_name;
+        new_param_sym.mode = ID_C;
```
No, make that the same mode as the function symbol's mode (which you look up in the next instruction).

peterschrammel replied:
The mode must be taken from the `function_name` symbol in the symbol table.

peterschrammel commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1448,6 +1449,32 @@ void goto_instrument_parse_optionst::instrument_goto_program()
+    status() << (only_generate ? "Generating" : "Replacing")
+             << " function bodies" << eom;
+    replace_function_bodies(
+      goto_model.goto_functions,
```
pass `goto_model` directly

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+  {
+    auto assume_instruction = function.body.add_instruction();
+    assume_instruction->make_assumption(false_exprt());
+    assume_instruction->code = code_assumet(false_exprt());
```
No, assume instructions should not have code set.

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+    assume_instruction->make_assumption(false_exprt());
+    assume_instruction->code = code_assumet(false_exprt());
+    auto end_instruction = function.body.add_instruction();
+    end_instruction->make_end_function();
```
Set up the `source_location` (to the location of the function symbol) and the `function` (to `function_name`) members of each instruction.

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+  {
+    auto assert_instruction = function.body.add_instruction();
+    assert_instruction->make_assertion(false_exprt());
+    assert_instruction->code = code_assertt(false_exprt());
```
Don't set the `code` member for an assertion.

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+    assert_instruction->source_location.set_comment(
+      irep_idt(
+        id2string(ID_assertion) + " " +
+        from_expr(ns, "", assert_instruction->guard)));
```
Use `format` instead of `from_expr`.

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+    auto assert_instruction = function.body.add_instruction();
+    assert_instruction->make_assertion(false_exprt());
+    assert_instruction->code = code_assertt(false_exprt());
+    assert_instruction->source_location.set_function(function_name);
```
See above: set `source_location` to the function symbol's location and set the `function` member.

peterschrammel replied:
Btw, who is responsible for calling `goto_functions.update()`?

hannes-steffenhagen-diffblue replied:
@peterschrammel The caller, in this case this happens in goto-instrument where update() is called after several transformations

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+    assume_instruction->make_assumption(false_exprt());
+    assume_instruction->code = code_assumet(false_exprt());
+    auto end_instruction = function.body.add_instruction();
+    end_instruction->make_end_function();
```
See above for several changes.

peterschrammel commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+  const std::regex &functions_regex,
+  bool only_generate,
+  const std::string &replace_options,
+  messaget &messages)
```
A `message_handlert` should be passed here.

peterschrammel commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+/// \param replace_options Options string indicating what body to generate
+/// \param messages Destination for status/warning messages
+void replace_function_bodies(
+  goto_functionst &functions,
```
Use `goto_modelt`

tautschnig commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+    const std::string &file,
+    int line,
+    const std::string &reason)
+    : runtime_error(file + ":" + std::to_string(line) + ": " + reason)
```
I would advise against making up your own error message. Users have no clue what's going on inside the code base, so give them an error message that refers to their input - no need to confuse them with cprover source file names.

peterschrammel commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+  virtual void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) = 0;
```
`const &` (several occurrences)

tautschnig commented on src/goto-instrument/replace_function_bodies.h:
```diff
@@ -0,0 +1,50 @@
+
+#include <regex>
+#include <vector>
+#include <iostream>
```
Neither `vector` nor `iostream` seem to be necessary here.

tautschnig reviewed (changes requested):
I think this is overall great work, but it does need a few tweaks as indicated by my comments. Almost all of these comments/changes need to be applied in several places beyond just the indicated ones.

peterschrammel commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+    symbol_tablet &symbol_table,
+    irep_idt function_name) override
+  {
+    auto assume_instruction = function.body.add_instruction();
```
How will stubbed functions be identifiable from information in the symbol table?

hannes-steffenhagen-diffblue replied:
What does that mean? Should stubbed functions be different from non-stubbed functions?

peterschrammel replied:
No, but for some applications (eg TG) we need to know whether a function has been stubbed.

peterschrammel commented on src/goto-instrument/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,388 @@
+public:
+  replace_function_bodies_errort(
+    const std::string &file,
+    int line,
```
why is this signed?

peterschrammel commented on src/goto-instrument/replace_function_bodies.h:
```diff
@@ -0,0 +1,50 @@
+#define OPT_GENERATE_FUNCTION_BODY "(generate-function-body):"
+#define OPT_REPLACE_FUNCTION_BODY_OPTIONS "(replace-function-body-options):"
+
+#define HELP_REPLACE_FUNCTION_BODY                                             \
```
I might have missed previous discussions, but why are these split up into three groups?

hannes-steffenhagen-diffblue replied:
Would you rather have the options and the help messages merged then?

peterschrammel replied:
I don't see a use case for only including one of these options into a tool.

peterschrammel commented on src/goto-instrument/replace_function_bodies.h:
```diff
@@ -0,0 +1,50 @@
+
+#define HELP_GENERATE_FUNCTION_BODY                                            \
+  " --generate-function-body <regex>\n"                                        \
+  "                              Like replace-function-body, but ignore "      \
```
Dangling reference to `replace-function-body` if `HELP_REPLACE_FUNCTION_BODY` is not included.

peterschrammel commented:
#1566 is subsumed by this PR and #1585 is not the way it should be done. Therefore, I suggest to close these two PRs.

hannes-steffenhagen-diffblue commented:
@peterschrammel @martin-cs @thk123 @chrisr-diffblue 

Things done in the interim:

* Moved the implementation of the functionality to goto-programs
* Moved the command line parsing into a reusable helper
* Added `const` in a bunch of places
* pass goto_modelt instead of goto_functionst/symbol_tablet pair
* use `format` instead of `from_expr` to convert expressions to string
* asserts/assumes no longer have code attached to them
* merged the OPT* and HELP* macros
* removed some unused includes
* use function mode for parameters instead of hardcoding to ID_C
* always set function/source_location on new instructions
* add null check before assigning pointer parameters

Outstanding:

* [x] Add additional regression tests for potentially confusing corner cases `int*` vs `int const*` vs `int * const` vs `int const * const` and `int const **`.
* [x] Clear up what should happen with parameters/globals that are structs with a mix of const and non-const members - ignore and print warning that global is being ignored because of const (__current behaviour__), or assign each member that is non-const
* [x] @tauschnig suggested removing replace-function-body entirely and only let generate-function-body generate bodies of functions that don't have one yet; If someone wanted the previous behaviour of replace they could use remove-function-body followed by generate-function-body instead; Please say if you also think this would be the right thing to do.
* [x] potentially remove file/line information from error messages - I personally dislike this, because someone who doesn't need the information can safely ignore it, but it could be useful for people with access to / understanding of the source code.

hannes-steffenhagen-diffblue commented:
@peterschrammel @thk123 @tautschnig @martin-cs 

Can you please check whether or not you feel that your comments have been addressed?

tautschnig commented on src/goto-programs/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,462 @@
+
+Module: Replace bodies of goto functions
+
+Author: DiffBlue Limited.
```
See #1919.

tautschnig commented on src/goto-programs/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,462 @@
+    };
+    auto assert_instruction = add_instruction();
+    assert_instruction->make_assertion(false_exprt());
+    // assert_instruction->code = code_assertt(false_exprt());
```
No need for this commented-out code, please remove.

tautschnig commented on src/goto-programs/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,462 @@
+  {
+    throw replace_function_bodies_errort(
+      "can only use one of --replace-function-body "
+      "or --generate-function-body");
```
@hannes-steffenhagen-diffblue @peterschrammel - did you discuss this? What was the conclusion, or maybe why was my proposal not acceptable?

hannes-steffenhagen-diffblue replied:
@tautschnig Nobody else said anything about this. It's possible that they missed it. I wouldn't mind implementing it that way, but it wasn't obviously better than the current behaviour from my perspective, I'd prefer at least one other person saying they think it's the better way to do it before going ahead with changing the options at this point.

tautschnig replied:
Poking @martin-cs @chrisr-diffblue with the hope that more people might chime in. Everyone: see the discussion in https://github.com/diffblue/cbmc/pull/1889#discussion_r175534313 for context.

chrisr-diffblue replied:
I've left a comment in that discussion thread - but bottom line is that I agree with @tautschnig's proposal.

tautschnig commented on src/goto-programs/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,462 @@
+/// A list of currently accepted command line arguments
+/// + the type of bodies generated by them
+///
+/// assert-false: { __CPROVER_assert(false); }
```
`__CPROVER_assert` takes a second argument.

hannes-steffenhagen-diffblue replied:
Ah. I should probably just use assert here, the __CPROVER doesn't really add to the legibility of the example

tautschnig commented on src/goto-programs/replace_function_bodies.h:
```diff
@@ -0,0 +1,118 @@
+
+Module: Replace bodies of goto functions
+
+Author: DiffBlue Ltd.
```
#1919

tautschnig commented on src/goto-programs/replace_function_bodies.h:
```diff
@@ -0,0 +1,118 @@
+#include <util/std_types.h>
+#include <util/message.h>
+#include <goto-programs/goto_model.h>
+#include <util/cmdline.h>
```
Please try to sort internal includes lexicographically.

tautschnig reviewed (commented):
Thanks a lot for the updates; still a few smaller issues, but mainly I'm wondering about the set options/the discussion around that, if it has taken place.

thk123 commented on regression/goto-instrument/replace-function-body-struct-with-const-member/main.c:
```diff
@@ -0,0 +1,21 @@
+  havoc_struct(&paramStruct);
+  assert(globalStruct.non_const == 10);
+  assert(globalStruct.is_const == 20);
+  assert(paramStruct.non_const == 11);
```
Why is this not been havoced? (A good technique for trying to verify these tests is to try and write bodies that violate this asserts, for example for this can you not do:

```C
(*WithConstMember).non_const = 20;
```

thk123 commented on src/goto-programs/replace_function_bodies.h:
```diff
@@ -0,0 +1,98 @@
+      const irep_idt &function_name);
+
+private:
+  /// Generate parameter names for unnamed parameters.
```
Comments should go with the implementation

thk123 commented on src/goto-programs/replace_function_bodies.cpp:
```diff
@@ -0,0 +1,410 @@
+/*******************************************************************\
```
I'm assuming this is a move of the original?

thk123 commented on regression/goto-instrument/replace-function-body-struct-with-const-member/main.c:
```diff
@@ -0,0 +1,21 @@
+{
+  struct WithConstMember paramStruct = {11, 21};
+  havoc_struct(&paramStruct);
+  assert(globalStruct.non_const == 10);
```
In fact - since you are also havocing globals, shouldn't this assert fail too?

thk123 commented on src/goto-programs/replace_function_bodies.cpp:
```diff
@@ -190,26 +194,35 @@ class havoc_replace_function_bodiest : public replace_function_bodiest
+        }
+        else if(!parameter.type().subtype().get_bool(ID_C_constant))
+        {
+          warning() << "warning: not havoccing non-const pointer parameter '"
```
Ah I see - this is why the tests are "wrong". I assume there is an issue? Probably worth ensuring this warning appears in the test and there is an appropriate comment on the issue to update the test.

hannes-steffenhagen-diffblue replied:
This is all intended behaviour for now, I had asked around what people thought the right behaviour should be and it was "just skip structs with const members for now". The other behaviour is also possible, would that be your preference then?

thk123 replied:
On the face of it it seems more "correct" as I can easily write a method body for that method that violates that assert without wandering into undefined behaviour or even something particularly odd. That said, I'm not using this tool on a regular basis, it might be in the real world this is simply a more useful behaviour.

thk123 reviewed (commented):
New changes seem fine but I think a bit more context on the slightly wrong test would be in order.

hannes-steffenhagen-diffblue commented:
@thk123 I've now changed the behaviour so that structs, unions and arrays are traversed during havoccing instead of giving up when they contain constant elements.

hannes-steffenhagen-diffblue commented:
There is some documentation about the feature in `doc/cprover-manual.md`. Unfortunately this has also lead to merge conflicts, which I will try to resolve now.

hannes-steffenhagen-diffblue commented:
I have rebased the changes onto the newest develop. Apologies for any inconveniences.

thk123 commented on regression/goto-instrument/replace-function-body-union-with-const-member/test.desc:
```diff
@@ -0,0 +1,14 @@
+^\[main.assertion.6\] assertion globalUnion.is_const == 10: FAILURE$
+^\[main.assertion.7\] assertion paramUnion.non_const == 20: FAILURE$
+^\[main.assertion.8\] assertion paramUnion.is_const == 20: FAILURE$
+^VERIFICATION FAILED$
```
Missing new line

thk123 reviewed (approved):
Thanks for changes and docs. I would suggest if the (?!__) thing is so important, you may want to add a warning in the code if the supplied regex matches `__CPROVER` since most people won't read the manual...

thk123 commented:
This is approaching the point where GH refuses to load this page due to too many comments. It might be wise to close this and open a new PR if there are still outstanding comments.

hannes-steffenhagen-diffblue commented:
@thk123 I might just end up doing that. Technical issues aside, it's starting to get to the point of being difficult to follow along with a lot of the discussion no longer being relevant due to code changes

tautschnig commented:
From my point of view this is very close to be ready to be merged (but others might disagree). I'd actually just suggest to squash the commit series into a few meaningful commits, re-review, approve, and be happy :-)

hannes-steffenhagen-diffblue commented:
@tautschnig I believe this is in a good state now. Would you mind taking another look at this once CI passes?

tautschnig commented on regression/goto-instrument/generate-function-body-union-with-const-member/test.desc:
```diff
@@ -0,0 +1,14 @@
+^\[main.assertion.6\] assertion globalUnion.is_const == 10: FAILURE$
+^\[main.assertion.7\] assertion paramUnion.non_const == 20: FAILURE$
+^\[main.assertion.8\] assertion paramUnion.is_const == 20: FAILURE$
+^VERIFICATION FAILED$
```
Github says there's no newline here.

hannes-steffenhagen-diffblue replied:
Fixed

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1522,9 +1535,10 @@ void goto_instrument_parse_optionst::help()
     " --check-invariant function   instruments invariant checking function\n"
     " --remove-pointers            converts pointer arithmetic to base+offset expressions\n" // NOLINT(*)
     " --splice-call caller,callee  prepends a call to callee in the body of caller\n"  // NOLINT(*)
+    " --undefined-function-is-assume-false\n"
```
Could we mark this as deprecated (possibly be removing it from help output, but leaving it in the .h file)? Also: is it now implemented via the new options? (Just being lazy, I should be able to figure out once I've completely read the code.)

hannes-steffenhagen-diffblue replied:
The effect can be replicated with `--generate-function-body '.*' --generate-function-body-options assume-false`

tautschnig replied:
Yes indeed - but then we could get rid of the code implementing `undefined_function_abort_path`, couldn't we?

tautschnig replied:
@hannes-steffenhagen-diffblue I'd leave it to you whether to include such changes in this PR or instead just create an issue and make someone else (such as myself...) fix it.

chrisr-diffblue replied:
I've created #2070 to cover the additional cleanup and will merge this PR now.

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.h:
```diff
@@ -86,7 +88,9 @@ Author: Daniel Kroening, kroening@kroening.com
   "(splice-call):" \
-  OPT_REMOVE_CALLS_NO_BODY
+  OPT_REMOVE_CALLS_NO_BODY \
+  OPT_REPLACE_FUNCTION_BODY
```
I think this might not be worth renaming to `OPT_GENERATE...`?

hannes-steffenhagen-diffblue replied:
done

tautschnig reviewed (approved):
I think there is further room for cleaning up the closely related options, but certainly this is a good piece of work to go in and further work can follow.

One wishlist request: please put a comment in the CHANGELOG file, this is a user-visible feature.

hannes-steffenhagen-diffblue commented:
@tautschnig I added a short reference to the new option in CHANGELOG under a new "5.9" section

chrisr-diffblue reviewed (approved):
Looks good - thanks for all the extra work and polishing on this PR.

polgreen commented:
I know this PR has already been merged, but I have a question about the documentation:

"  /// Replace the function body with one based on the implementation
  /// This will work the same whether or not the function already has a body"

Does "implementation" have a specific meaning, I thought it just meant the actual function body? I'm not sure how you would replace the function body with one based on the function body if there is no function body, or, if there is a body, how the body that it replaces the body with is based on the original body?
